### PR TITLE
Populate document outline with all symbols, not just top level

### PR DESCRIPTION
**Pull request by MikhailArkhipov**
https://github.com/Microsoft/vscode-python/issues/2334
https://github.com/Microsoft/vscode-python/issues/2050

### CODE REVIEW

File: Python/Product/Analysis/LanguageServer/Server.WorkspaceSymbols.cs [hunk 1]
```diff
@@ -40,30 +42,61 @@ await GetModuleVariablesAsync(entry as ProjectEntry, opts, @params.query)
...
+                    yield return m;
                 }
             }
         }
 
+        private static IEnumerable<MemberResult> GetScopeVariables(ModuleAnalysis analysis, InterpreterScope scope, GetMemberOptions opts) {
+            var res = analysis.GetAllAvailableMembersFromScope(scope, opts).Concat(GetChildScopesVariables(analysis, scope, opts));
+            foreach (var m in res) {
+                yield return m;
+            }
```
Comment (zooba) [inline Python/Product/Analysis/LanguageServer/Server.WorkspaceSymbols.cs]: `return res`?

File: Python/Product/Analysis/LanguageServer/Server.WorkspaceSymbols.cs [hunk 2]
```diff
@@ -40,30 +42,61 @@ await GetModuleVariablesAsync(entry as ProjectEntry, opts, @params.query)
...
+            .Concat(GetChildScopesVariables(analysis, analysis.Scope, opts));
+        }
+
+        private static IEnumerable<MemberResult> GetChildScopesVariables(ModuleAnalysis analysis, InterpreterScope scope, GetMemberOptions opts) {
+            foreach (var childScope in scope.Children) {
+                var res = GetScopeVariables(analysis, childScope, opts);
+                foreach (var m in res) {
+                    yield return m;
                 }
             }
```
Comment (zooba) [inline Python/Product/Analysis/LanguageServer/Server.WorkspaceSymbols.cs]: `return scope.Children.SelectMany(c => GetScopeVariables(analysis, c, opts))`?

File: Python/Product/Analysis/ModuleAnalysis.cs [hunk 3]
```diff
@@ -710,6 +692,40 @@ private class ErrorWalker : PythonWalker {
...
+                if (scopeResult.TryGetValue(kvp.Key, out var values)) {
+                    values.AddRange(vars);
+                } else {
+                    scopeResult[kvp.Key] = values = new List<AnalysisValue>(vars);
+                }
+                values.Add(new SyntheticDefinitionInfo(
+                    kvp.Key,
+                    null,
+                    kvp.Value.Definitions.Select(e => e.GetLocationInfo())
+                ));
```
Comment (zooba) [inline Python/Product/Analysis/ModuleAnalysis.cs]: Do we need this if there have been variables added? Or do we need it if the value has no definitions?

File: Python/Product/Analysis/Values/BuiltinNamespace.cs [hunk 4]
```diff
@@ -133,6 +133,6 @@ class BuiltinNamespace<MemberContainerType> : AnalysisValue where MemberContaine
             return false;
         }
 
-        public override int GetHashCode() => new { hc1 = GetType().GetHashCode(), hc2 = _type.GetHashCode() }.GetHashCode();
+        public override int GetHashCode() => new { hc1 = GetType()?.GetHashCode() ?? 0, hc2 = _type?.GetHashCode() ?? 0 }.GetHashCode();
```
Comment (zooba) [inline Python/Product/Analysis/Values/BuiltinNamespace.cs]: Think I just saw a slightly different fix in another PR, but now I look at it, we shouldn't have to call `GetHashCode()` on the members here. The anonymous type `GetHashCode` will do it (and handle nulls properly)

File: Python/Product/Analysis/LanguageServer/Server.WorkspaceSymbols.cs (hunk 2 above)
Comment (MikhailArkhipov) [reply Python/Product/Analysis/LanguageServer/Server.WorkspaceSymbols.cs]: Yep :-)

File: Python/Product/Analysis/Values/BuiltinNamespace.cs (hunk 4 above)
Comment (MikhailArkhipov) [reply Python/Product/Analysis/Values/BuiltinNamespace.cs]: Yes, will merge

File: Python/Product/Analysis/ModuleAnalysis.cs (hunk 3 above)
Comment (MikhailArkhipov) [reply Python/Product/Analysis/ModuleAnalysis.cs]: It was existing code that applied to a single scope and now called for each scope recursively.

**Pull request merged**